### PR TITLE
Add parallel edge graph support

### PR DESCRIPTION
**Pull request by ghsteff**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Edges coming from the same source and going to the same target (called parallel edges in graphology docs) aren't being rendered when multiple of them exist

## What is the new behavior?
Parallel edges are all displayed

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
Screenshot 2023-12-07 at 12 54 58 PM
Screenshot 2023-12-07 at 1 08 07 PM

https://github.com/reaviz/reagraph/assets/40581813/059888ef-255f-4208-afa6-b539af14efc3

### CODE REVIEW

File: src/symbols/Edge.tsx [hunk 1]
```diff
@@ -99,6 +99,8 @@ export interface EdgeProps {
   onPointerOut?: (edge: InternalGraphEdge) => void;
 }
 
+const MULTI_EDGE_OFFSET_FACTOR = 0.7;
```
Comment (ghsteff) [inline src/symbols/Edge.tsx:102]: Could add this as a prop somewhere @amcdnl ? Not sure if it's something ppl want to customize and I didn't wanna just add extra noise to the prop list

Here's what it does 
image

image

File: src/store.ts [hunk 2]
```diff
@@ -73,7 +73,7 @@ export const createStore = ({
     edgeMeshes: [],
     selections,
     drags: {},
-    graph: new Graph(),
+    graph: new Graph({ multi: true }),
```
Comment (amcdnl) [inline src/store.ts:76]: Whats this do?

File: src/symbols/Edge.tsx (hunk 1 above)
Comment (amcdnl) [reply src/symbols/Edge.tsx:102]: Ya I wouldn't expose for now

File: src/symbols/Edge.tsx [hunk 3]
```diff
@@ -114,36 +116,73 @@ export const Edge: FC<EdgeProps> = ({
...
   const draggingId = useStore(state => state.draggingId);
   const [active, setActive] = useState<boolean>(false);
   const [menuVisible, setMenuVisible] = useState<boolean>(false);
-  const labelOffset = (size + theme.edge.label.fontSize) / 2;
+  let labelOffset = (size + theme.edge.label.fontSize) / 2;
+  let curveOffset: number;
 
   const [arrowLength, arrowSize] = useMemo(() => getArrowSize(size), [size]);
 
+  if (edgesWithSameSourceAndTarget.length > 1) {
```
Comment (amcdnl) [inline src/symbols/Edge.tsx:137]: This needs some comments - I might even move it to a pure function for easier isolation.

File: src/symbols/Edge.tsx [hunk 4]
```diff
@@ -155,6 +194,14 @@ export const Edge: FC<EdgeProps> = ({
     if (curved) {
       // Offset the label to the mid point of the curve
       const offset = new Vector3().subVectors(newMidPoint, curve.getPoint(0.5));
+      switch (labelPlacement) {
+      case 'above':
+        offset.y = offset.y - 3;
```
Comment (amcdnl) [inline src/symbols/Edge.tsx:199]: I'd make this `3` a const

File: src/store.ts (hunk 2 above)
Comment (amcdnl) [reply src/store.ts:76]: Probably for cycle detection on certain layouts. Probably fine for now though just check all the layouts work as expected

**Comment by peppekens** [comment]
why can't I find these changes in the npm package?

**Pull request merged**